text/0000-multi-context-module.md

```diff
@@ -0,0 +1,200 @@
+- Start Date: 2020-08-30
+- RFC PR: (leave this empty)
+- Svelte Issue: (leave this empty)
+
+# Multiple `context=module` Scripts
+
+## Summary
+
+Allow for multiple `script[context=module]` blocks to exist, targetting different output/generate modes.
+
+In general, this will (probably) be used _primarily_ by Sapper (and its relatives) but, the ability must be implemented by Svelte. More importantly, this makes sense for Svelte in its own right, too, since a flagship feature of Svelte is that it can compile/generate its own `ssr`-vs-`dom` outputs.
+
+## Motivation
+
+In many – but not most – situations, it makes sense to export different module methods depending on _how_ a component is compiled. The most popular example is `preload` and so I'll use that as [an example](#example), but this RFC is most definitely not limited to nor specifically targetting `preload` alone.
+
+## Detailed Design
+
+By default, `context=module` will act the same as it does now. This is because there's no reason for it to be any different - which also means that this would be a non-breaking change/addition.
+
+While syntax is still WIP and up for debate, I'm thinking something like these will work well:
+
+#### Option 1
+
+```html
+<script context="module">
+    <!-- shared with "dom" and "ssr" -->
+</script>
+
+<script context="module:dom">
+    <!-- added to "dom" output only -->
+</script>
+
+<script context="module:ssr">
+    <!-- added to "ssr" output only -->
+</script>
+```
+
+#### Option 2
+
+> **Note:** I am avoiding `type=` intentionally to avoid IDE/tooling issues.
+
+```html
+<script context="module">
+    <!-- shared with "dom" and "ssr" -->
+</script>
+
+<script context="module" output="dom">
+    <!-- added to "dom" output only -->
+</script>
+
+<script context="module" output="ssr">
+    <!-- added to "ssr" output only -->
+</script>
+```
+
+I lean towards Option 1.
+
+
+## Example
+
+> The thought experiment may be more familiar under the guise of Sapper, but it's not **for** Sapper specifically.
+
+Let's assume you have an `Article` component. It's self-reliant, meaning it fetches its own data (based on a `slug` prop, perhaps) and renders the article body with the given response:
+
+```html
+<script lang="ts">
+    let post: Article = {};
+
+    export let slug: string;
+</script>
+
+<h1>{post.title}</h1>
+
+<div class="content">
+	{@html post.body}
+</div>
+```
+
+Generally, something like this warrants a `preload` function inside a `context=module` block, and then a runtime (Sapper, or otherwise) will call this `preload` function so that the `Article` instance has its data handy at time of render. We'll use `fetch` for this:
+
+```html
+<script context="module" lang="ts">
+    let loaded = {};
+
+    export function preload(req: IRequest) {
+        loaded.slug = req.params.slug;
+        return fetch(`.../${loaded.slug}`).then(r => r.json()).then(obj => {
```

Comment:
should this be `this.fetch`?

(also, there's a mix of tabs and spaces here that makes it slightly harder to read)

Reply:
No, this really has nothing to do with Sapper. Here, `fetch()` is intentional as it's a browser API and would throw an error/require polyfill on server-side. The premise of the example is that a developer wants to implement a non-`fetch` preloader for `ssr` outputs.

Reply:
I did this with the new GitHub codespaces and didn't notice any space inconsistencies. Will go thru now and try to fix